bundles/org.openhab.binding.insteon/src/main/java/org/openhab/binding/insteon/internal/InsteonBinding.java

```diff
@@ -0,0 +1,517 @@
+    private int m_deviceStatisticsInterval = 600; // in seconds
+    private int m_messagesReceived = 0;
+    private boolean m_isActive = false; // state of binding
+    private int m_x10HouseUnit = -1;
+    private InsteonNetworkHandler m_handler;
+
+    public InsteonBinding(InsteonNetworkHandler handler, String port, @Nullable Integer deviceStatisticsInterval,
+            @Nullable Integer deviceDeadCount, @Nullable Integer devicePollInterval, @Nullable String additionalDevices,
+            @Nullable String additionalFeatures, @Nullable Integer modemDbRetryTimeout) {
+
+        this.m_handler = handler;
+
+        if (deviceStatisticsInterval != null) {
+            m_deviceStatisticsInterval = deviceStatisticsInterval;
+            logger.info("device statistics interval set to {} seconds", m_deviceStatisticsInterval);
+        }
+
+        int deadDeviceCount = 10;
+        if (deviceDeadCount != null) {
+            deadDeviceCount = deviceDeadCount;
+            logger.info("device dead co- Initial contributionunt set to {}", deadDeviceCount);
+        }
+
+        if (devicePollInterval != null) {
+            m_devicePollInterval = devicePollInterval;
+            logger.info("device poll interval set to {} milliseconds", m_devicePollInterval);
+        }
+
+        if (additionalDevices != null) {
+            try {
+                DeviceTypeLoader.s_instance().loadDeviceTypesXML(additionalDevices);
+                logger.info("read additional device definitions from {}", additionalDevices);
+            } catch (Exception e) {
+                logger.error("error reading additional devices from {}", additionalDevices, e);
+            }
+        }
+
+        if (additionalFeatures != null) {
+            logger.info("reading additional feature templates from {}", additionalFeatures);
+            DeviceFeature.s_readFeatureTemplates(additionalFeatures);
+        }
+
+        if (modemDbRetryTimeout != null) {
+            int timeout = modemDbRetryTimeout;
+            m_driver.setModemDBRetryTimeout(timeout * 1000);
+            logger.info("setting modem db retry timeout to {} seconds", timeout);
+        }
+
+        m_deadDeviceTimeout = m_devicePollInterval * deadDeviceCount;
+        logger.info("dead device timeout set to {} seconds", m_deadDeviceTimeout / 1000);
+
+        logger.info("port = '{}'", port);
+        m_driver.addPort("port", port);
+        m_driver.addMsgListener(m_portListener, port);
+
+        logger.debug("setting driver listener");
+        m_driver.setDriverListener(m_portListener);
+    }
+
+    public boolean startPollingWhenReady() {
+        // hold off on starting to poll until devices have been added From the Things file
+        int settleTime = 5000;
+        for (int i = -1; i != m_devices.size(); i = m_devices.size()) {
+            try {
+                logger.debug("sleeping {} milliseconds...", settleTime);
+                Thread.sleep(settleTime);
```

Comment:
This sleep is really bad. The scheduled threads of all bindings use a shared thread pool. If you sleep for 5 seconds it means you block a thread to be used by other bindings for 5 seconds. You should reschedule here to start after 5 seconds.

Reply:
This method is called from a scheduler.execute block within InsteonNetworkHandler.Initialize() and this method needs to block until it can poll, any suggestions on how to make this happen?